### ISSUE TITLE
N999 false positive for a directory named as a dotted module path (e.g. `coherent.deps`)

### ISSUE THREAD

**Issue by jaraco** [astral-sh/ruff#27801]
## Summary

`ruff check` reports `N999 Invalid module name: 'coherent.deps'` for a package directory named `coherent.deps/` — under **default rules**, with no configuration and no `import` of the package anywhere. But `coherent.deps` is a **valid** (dotted) module name: it's exactly the name under which the package is imported and appears in `sys.modules`.

## Reproduce (ruff 0.16.3)

```console
$ mkdir coherent.deps
$ printf '"""A package."""\n' > coherent.deps/__init__.py
$ ruff check coherent.deps/
coherent.deps/__init__.py:1:1: N999 Invalid module name: 'coherent.deps'
Found 1 error.
```

## Two surprises

1. **Nothing imports it.** There is no `import coherent.deps` anywhere — the package isn't imported at all. N999 fires purely from the file's directory name, on the default rule set. That makes the diagnostic hard to attribute: the "module" being flagged is the file currently under lint, not something referenced by an import.

2. **`coherent.deps` is a valid module name.** It is the dotted name in `sys.modules`. N999 appears to treat the single path component `coherent.deps` as a *flat* module name and reject the embedded `.`, but interpreted as a dotted module **path** — which is how the package is imported — it is perfectly valid. The message ("Invalid module name") is incorrect on its face.

## Context

This directory layout is used throughout [coherent-oss](https://github.com/coherent-oss): the source directory is named for the **distribution** (`coherent.deps`, `compilers.C`, …) and is redirected to the namespaced **import package** (`coherent/deps/`, `compilers/C/`) at build/install time. So the physical directory name is intentionally the dotted distribution/module name.

This is the same "redirected / essential layout" that ty is working to support — astral-sh/ty#3603 (essential layout), astral-sh/ty#475 (setuptools-style editable installs), with a proposed fix in #25984 (honor `import redirect` comments in `.pth` files). Since N999 is a **ruff lint rule** (distinct from ty's module resolution), it may or may not share that fix — but note that even with no redirection concept at all, `coherent.deps` is a valid dotted module path and arguably shouldn't be flagged.

## Suggestion

N999 should not report `Invalid module name` for a directory whose name is a valid **dotted module path** (treat `.` as a package separator), and/or should account for namespace/redirected layouts.

Current workaround: ignore `N999` in the ruff config.

**Comment by ntBre** [comment]
I might be missing something here, but this seems to be working as intended. Dots are not valid in module names, they're valid in module paths:

```console
❯ mkdir foo.bar
❯ echo 'class C: ...' > foo.bar/__init__.py
❯ python -c 'from foo.bar import C'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    from foo.bar import C
ModuleNotFoundError: No module named 'foo'
```

That import would be valid for `foo/bar/__init__.py`, not `foo.bar/__init__.py`.

Ruff is a single-file linter, so yes, it's only checking the module name, not searching for files that actually import it. And the usage in `coherent-oss` seems unusual enough that ignoring the rule seems like the right answer.

Based on the responses on https://github.com/astral-sh/ty/issues/3603, I think it's a bit of a stretch to say that ty is working to support this layout, and I think we'd need to see more demand for it to try supporting it in Ruff too.
